### PR TITLE
Include ContentLength when enumerating HttpHeaders

### LINKED ISSUE #54557: Net7 regression: Kestrel HttpRequestHeaders doesn't enumerate "Content-Length" unless there's another known header also present.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

If you add a key to a Dictionary, then enumerate its elements just afterwards, you expect the recently added key to be present.

However this is not the case for the `Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestHeaders` class starting from net7.0, if the key is `"Content-Length"`, and the dictionary does not contain any other known headers:
``` C#
httpRequestHeaders["Content-Length"] = "423";
        - Count: 1
        - Keys: []
        - Values: []
        - Enumerate: []
        - TryGetValue("Content-Length", out var _) == True
```

E.g. some of the properties on the directionary act as if the `"Content-Length"` was added, others act as if it wasn't.

What's interesting to is that if there is at least one other known header in the dictionary, regardless if it was added before or after the `"Content-Length"`, then the behaviour of the dictionary becomes normal again.

E.g. for the above example, adding `"Content-Type"` to the headers makes the `"Content-Length"` appear in the enumerated list as well:
``` C#
httpRequestHeaders["Content-Type"] = "text";
        - Count: 2
        - Keys: [Content-Type, Content-Length]
        - Values: [text, 423]
        - Enumerate: [[Content-Type, text], [Content-Length, 423]]
        - TryGetValue("Content-Length", out var _) == True 
```

### Expected Behavior

Since the `HttpRequestHeaders` implements the `IDictionary` interface, it should behave like a `Dictionary` in respect to adding and enumarating elements (e.g. how it behaved until `net6.0`), and the number of enumerated elements should always be equal to the `.Count` property.

Even if `"Content-Length"` is treated in a special way, the presence of other known headers should not impact whether `"Content-Length"` is enumerated or not.

### Steps To Reproduce

Create a console app:

## The .csproj:
```xml
<Project Sdk="Microsoft.NET.Sdk.Web">
    <PropertyGroup>
        <TargetFramework>net8.0</TargetFramework>
        <OutputType>Exe</OutputType>
        <ImplicitUsings>enable</ImplicitUsings>
    </PropertyGroup>
</Project>
```

## The Program.cs:
``` C#
using System.Text;
using Microsoft.Net.Http.Headers;

public static class Program
{
    public static void Main()
    {
        IHeaderDictionary dict =
            (IHeaderDictionary)
            typeof(Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServer)
            .Assembly
            .GetType("Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestHeaders", throwOnError: true)!
            .GetConstructor([typeof(bool), typeof(Func<string, Encoding>)])!
            .Invoke([true, null]);

        AddAndPrint(dict, "a", "1");
        AddAndPrint(dict, HeaderNames.ContentLength, "423");
        AddAndPrint(dict, "b", "2");
        AddAndPrint(dict, HeaderNames.ContentType, "text");
    }

    private static void AddAndPrint(IHeaderDictionary dict, string key, string value)
    {
        Console.WriteLine($"httpRequestHeaders[\"{key}\"] = \"{value}\";");
        dict[key] = value;
        Console.WriteLine("\t- Count: " + dict.Count);
        Console.WriteLine("\t- Keys: [" + string.Join(", ", dict.Keys) + "]");
        Console.WriteLine("\t- Values: [" + string.Join(", ", dict.Values) + "]");
        Console.WriteLine("\t- Enumerate: [" + string.Join(", ", dict) + "]");
        Console.WriteLine($"\t- TryGetValue(\"{HeaderNames.ContentLength}\", out var _) == " + dict.TryGetValue(HeaderNames.ContentLength, out var x) + "\n");
    }
}


```

### Exceptions (if any)

_No response_

### .NET Version

7.0.0 or newer

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by amcasey**
...even if no bits are set.

Fixes #54557

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestHeaders.cs [hunk 1]
```diff
@@ -263,6 +263,7 @@ internal Enumerator(HttpRequestHeaders collection)
         public void Reset()
         {
             _currentBits = _collection._bits;
+            _next = GetNext(_currentBits, _collection.ContentLength.HasValue);
```
Comment (BrennanConroy) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestHeaders.cs:266]: This gets overwritten on the next line.

File: src/Servers/Kestrel/shared/KnownHeaders.cs [hunk 2]
```diff
@@ -1428,7 +1428,16 @@ public bool MoveNext()
                     {(!loop.ClassName.Contains("Trailers") ? $@"_next = _collection._contentLength.HasValue ? {loop.Headers.Length - 1} : -1;" : "_next = -1;")}
                     return true;
                 }}
-            }}
+            }}{(loop.ClassName.Contains("Trailers") ? "" : $@"
+
+            private static int GetNext(long bits, bool hasContentLength)
```
Comment (BrennanConroy) [inline src/Servers/Kestrel/shared/KnownHeaders.cs:1433]: Should this have `[MethodImpl(MethodImplOptions.AggressiveInlining)]`

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestHeaders.cs [hunk 3]
```diff
@@ -263,6 +263,7 @@ internal Enumerator(HttpRequestHeaders collection)
...
             _next = _currentBits != 0 ? BitOperations.TrailingZeroCount(_currentBits) : -1;
```
Comment (amcasey) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestHeaders.cs:267]:
```suggestion
```

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestHeaders.cs (hunk 1 above)
Comment (amcasey) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestHeaders.cs:266]: Good catch

File: src/Servers/Kestrel/shared/KnownHeaders.cs (hunk 2 above)
Comment (amcasey) [reply src/Servers/Kestrel/shared/KnownHeaders.cs:1433]: You tell me.  When do we apply that?

Comment (BrennanConroy) [reply src/Servers/Kestrel/shared/KnownHeaders.cs:1433]: When it shouldn't be a function call 😆 
This function is small enough that we would be fine with the code bloat in favor of no function call overhead.

**Review by BrennanConroy — APPROVED** [review]

**Pull request merged**